### PR TITLE
Add Schedule-Free Signum (Sign SGD) optimizer 

### PR DESCRIPTION
**Pull request by luciaquirke**
https://github.com/EleutherAI/sparsify/pull/54 with legacy Adam + linear lr schedule optionally available.

> It turns out that [Sign SGD](https://arxiv.org/abs/1802.04434) (simply descending the elementwise sign of the stochastic gradient) is an amazing optimizer for training SAEs and transcoders. It dramatically reduces the number of dead neurons in cases where that's a problem with Adam, presumably because it aggressively magnifies the updates applied to infrequent features, while clipping the update applied to frequent features for enhanced training stability. Convergence is greatly accelerated and often ends up at a lower final FVU.

> Adding a [schedule-free wrapper](https://arxiv.org/abs/2405.15682) makes it even better, both in terms of convergence speed and dead neuron reduction. We also quadruple the default batch size, since that seems to enhance final FVU and reduce dead neurons, especially with the new optimizer.

### CODE REVIEW

File: sparsify/config.py [hunk 1]
```diff
@@ -54,11 +54,15 @@ class TrainConfig(Serializable):
     micro_acc_steps: int = 1
     """Chunk the activations into this number of microbatches for training."""
 
+    optimizer: Literal["adamw", "signum"] = "signum"
```
Comment (norabelrose) [inline sparsify/config.py:57]: let's call it Adam because we never use weight decay

File: sparsify/trainer.py [hunk 2]
```diff
@@ -82,43 +82,57 @@ def __init__(
...
-            from torch.optim import Adam
+                print("bitsandbytes 8-bit Adam not available, using torch.optim.Adam")
+                print("Run `pip install bitsandbytes` for less memory usage.")
 
-            print("bitsandbytes 8-bit Adam not available, using torch.optim.Adam")
-            print("Run `pip install bitsandbytes` for less memory usage.")
+            assert isinstance(dataset, Sized)
+            num_examples = len(dataset)
 
+            self.optimizer = Adam(pgs)
```
Comment (norabelrose) [inline sparsify/trainer.py:116]: yeah we're just not using adamw

**Review by norabelrose — APPROVED** [review]
Just one nitpick
